File: SafetyProgram.UI/MainWindowView/MainViewModel.cs
using System.Collections.Generic;
using System.ComponentModel;
using System.Windows.Controls;
using System.Windows.Input;
using Fluent;
using SafetyProgram.Base;
using SafetyProgram.Base.Interfaces;
using SafetyProgram.Core.Commands;
using SafetyProgram.UI.Document;

namespace SafetyProgram.UI
{
    public sealed class MainViewModel : 
        IMainViewModel
    {
        public MainViewModel(Ribbon ribbonView,
            Control contentView)
        {
            Helpers.NullCheck(ribbonView,
                contentView);

            this.RibbonView = ribbonView;
            this.ContentView = contentView;
        }

        public Ribbon RibbonView { get; private set; }

        public Control ContentView { get; private set; }

        public event PropertyChangedEventHandler PropertyChanged;
    }
}

File: SafetyProgram.UI.ViewModels/MainViewModel.fs
namespace SafetyProgram.UI.ViewModels

open SafetyProgram.UI.Models
open System.Windows.Controls
open SafetyProgram.Base.Helpers
open System
open Fluent
open FSharpx.Option

type MainViewModel(ribbonVVM : GuiKernelData -> Ribbon * IViewModel, contentVVM : Option<GuiDocument> -> Control * IViewModel, model : GuiKernelData) as this =

    let propertyChanged = new Event<_,_>()

    let contentViewGen x =
        model.Content
        >>= fun y -> Some y.Content
        |> contentVVM

    let ribbonView, ribbonViewModel = ribbonVVM model
    let mutable contentView, contentViewModel = contentViewGen model.Content

    let commands = new MainCommands(model)
    
    let handler = model.ContentChanged.Subscribe(fun newContent ->
        let x, y = contentViewGen newContent
        contentView<-x
        contentViewModel<-y
        raisePropChanged propertyChanged this "ContentView")

    interface IMainViewModel with

        member this.RibbonView = ribbonView
        member this.ContentView = contentView

        [<CLIEvent>]
        member this.PropertyChanged = propertyChanged.Publish

        member this.Dispose () =
            ribbonViewModel.Dispose()
            contentViewModel.Dispose()
            handler.Dispose()
            (commands :> IDisposable).Dispose()

        member this.Commands = commands

    // IMPLICIT IMPL
    member this.RibbonView = ribbonView
    member this.ContentView = contentView

    [<CLIEvent>]
    member this.PropertyChanged = propertyChanged.Publish

    member this.Dispose () =
        ribbonViewModel.Dispose()
        contentViewModel.Dispose()
        handler.Dispose()
        (commands :> IDisposable).Dispose()

    member this.Commands = commands